### ISSUE TITLE
Vendoring dune repositories lead to build failure because of a blackbox test

### ISSUE THREAD

**Issue by NathanReb** [ocaml/dune#4001]
Vendoring dune.2.7.1's repository in a project, for example when using `opam-monorepo` while depending on `dune-build-info`, leads to the following error whenever one tries to build anything:
```
$ dune build
File "dune_/test/blackbox-tests/test-cases/dialects.t/bad1/dune-project", line 9, characters 1-74:
 9 |  (name d)
10 |  (implementation (extension foo2))
11 |  (interface (extension bar2)))
Error: dialect "d" is already defined
```

This happens when one's running dune.2.7.1 themselves. To reproduce, simply run:
```
echo "(lang dune 2.7)" > dune-project
echo "(vendored_dirs dune_)" > dune
git clone https://github.com/ocaml/dune.git -b 2.7.1 dune_
dune build
```

I know this is a bit of a corner case but I'm under the impression that cram tests directories shouldn't pollute the dune-workspace that way, especially since we're not trying to run them here. Is this behaviour expected or should it be considered a bug?

**Comment by emillon** [comment]
I was looking for a quick workaround and convert the dialect test to a cram file (where we write dune-project within the test itself). I started doing that, but there are a lot more cases than dialects. It seems that all test cases that make dune fail make the outer dune fail too.

**Comment by emillon** [comment]
No big surprises but I bisected and this behaviour is indeed introduced by 48406c09fc7d8d9b91ad2daf8d2d4568fe0d4c41.

**Comment by emillon** [comment]
One relevant bit is that test-cases used to be marked a data_only_dir in a dune stanza and it's not done in this piece of code:

https://github.com/ocaml/dune/blob/eca6071b4c1222107d6d10da1fbd26c14c2cfeab/src/dune_engine/file_tree.ml#L593-L602

It might not work in the case where the cram directory is not the immediate parent, I'll check that.

**Comment by rgrinberg** [comment]
Didn't we fix this in #3860 already?

**Comment by emillon** [comment]
Oh, indeed that works. I was only bisecting by changing vendored dune sources, but it's the outer dune that needed to be fixed. Closing then. opam-monorepo users will love the next release :+1:
